sdk/azcore/arm/resource_identifier.go

```diff
@@ -0,0 +1,315 @@
+//go:build go1.16
+// +build go1.16
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package arm
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+)
+
+const (
+	providersKey             = "providers"
+	subscriptionsKey         = "subscriptions"
+	resourceGroupsLowerKey   = "resourcegroups"
+	locationsKey             = "locations"
+	builtInResourceNamespace = "Microsoft.Resources"
+)
+
+var (
+	RootResourceIdentifier = &ResourceIdentifier{
+		parent:       nil,
+		resourceType: TenantResourceType,
+		name:         "",
+	}
+
+	SubscriptionResourceType  = NewResourceType("Microsoft.Resources", "subscriptions")
+	ResourceGroupResourceType = NewResourceType("Microsoft.Resources", "resourceGroups")
+	TenantResourceType        = NewResourceType("Microsoft.Resources", "tenants")
+	ProvidersResourceType     = NewResourceType("Microsoft.Resources", "providers")
+)
+
+// ResourceIdentifier represents a resource ID
+type ResourceIdentifier struct {
+	parent            *ResourceIdentifier
+	subscriptionId    string
+	provider          string
+	resourceGroupName string
+	location          string
+	resourceType      ResourceType
+	name              string
+	isChild           bool
+
+	stringValue *string
```

Comment:
Does this really need to be a pointer or is the empty string good enough.

Reply:
Make sense. And serialized resource identifier can never be empty, therefore I changed this to `string`